vowpalwabbit/core/src/reductions/cb/cb_explore_adf_graph_feedback.cc

```diff
@@ -0,0 +1,527 @@
+#include "vw/config/options.h"
+#include "vw/core/gd_predict.h"
+#include "vw/core/global_data.h"
+#include "vw/core/label_dictionary.h"
+#include "vw/core/label_parser.h"
+#include "vw/core/model_utils.h"
+#include "vw/core/parser.h"
+#include "vw/core/reductions/cb/cb_adf.h"
+#include "vw/core/reductions/cb/cb_explore.h"
+#include "vw/core/reductions/cb/cb_explore_adf_common.h"
+#include "vw/core/setup_base.h"
+#include "vw/core/vw_math.h"
+#include "vw/explore/explore.h"
+
+#include <algorithm>
+#define ARMA_DONT_USE_BLAS
+#define ARMA_DONT_USE_LAPACK
+#include <armadillo>
+#include <cmath>
+#include <ensmallen.hpp>
+#include <functional>
+#include <vector>
+
+using namespace VW::cb_explore_adf;
+
+namespace VW
+{
+namespace cb_explore_adf
+{
+class cb_explore_adf_graph_feedback
+{
+public:
+  cb_explore_adf_graph_feedback(float gamma, VW::workspace* all) : _gamma(gamma), _all(all) {}
+  // Should be called through cb_explore_adf_base for pre/post-processing
+  void predict(VW::LEARNER::learner& base, multi_ex& examples);
+  void learn(VW::LEARNER::learner& base, multi_ex& examples);
+  void save_load(io_buf& io, bool read, bool text);
+  size_t _counter = 0;
+  float _gamma;
+
+private:
+  VW::workspace* _all;
+  template <bool is_learn>
+  void predict_or_learn_impl(VW::LEARNER::learner& base, multi_ex& examples);
+  void update_example_prediction(multi_ex& examples);
+};
+}  // namespace cb_explore_adf
+
+/**
+ * Implementing the constrained optimization from the paper: https://arxiv.org/abs/2302.08631
+ * We want to minimize objective: p * fhat + z
+ * where p is the resulting probability distribution, fhat is the scores we get from cb_adf and z is a constant
+ * The constaints are (very roughly) for each action we want:
+ *        (p - ea)**2 / G * p <= fhat_a + z
+ * where fhat_a is the cost for that action and ea is the vector of the identity matrix at index a
+ *
+ * G is num_actions x num_actions graph where each column indicates the probability of that action (corresponding to the
+ * column index) revealing the reward for a different action (which row are we talking about)
+ *
+ * So for example the identity matrix for a 2 x 2 matrix
+ * 1 0
+ * 0 1
+ *
+ * basically means that each action taken will reveal information only for itself
+ *
+ * and the all 1's matrix
+ *
+ * 1 1
+ * 1 1
+ *
+ * corresponds to supervised learning where each action taken will reveal infromation for every other action
+ *
+ *
+ * The way to think about gamma, which will increase over time, is that a small gamma means that the final p will
+ * "listen" more to the graph and try to give higher probability to the actions that reveal more, even if they have a
+ * higher cost. As gamma increases, the final p will "listen" less to the graph and will listen more to the cost, so if
+ * an action has a high cost it will have a lower probability of being chosen even if it reveals a lot of information
+ * about other actions
+ */
+class ConstrainedFunctionType
```

Comment:
Update style in this file to match the repo

Reply:
I can change the class style but the functions need to stay as they are since they will be called by the AugLagrangian optimizer, so it will look nicer if the class name follows the rest of the class style imo